File: src/main/java/io/myosotisdev/minestom/command/AbstractCommand.java
package io.myosotisdev.minestom.command;

import io.myosotisdev.minestom.permission.Permission;
import io.myosotisdev.minestom.util.component.Components;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
import net.minestom.server.command.builder.Command;
import net.minestom.server.command.builder.arguments.Argument;
import net.minestom.server.command.builder.arguments.ArgumentEnum;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Iterator;

public abstract class AbstractCommand extends Command
{
    private AbstractCommand parent;

    public AbstractCommand(AbstractCommand command, @NotNull String name, @Nullable String... aliases)
    {
        super(name, aliases);

        this.parent = command;

        setDefaultExecutor((sender, context) -> {
            sender.sendMessage(Components.fromLegacy("&aCommand: &8[&6" + getName() + "&8]"));

            sender.sendMessage(Components.fromLegacy("&7Aliases:"));
            for (String alias : getAliases())
            {
                sender.sendMessage(Components.fromLegacy("&7 - &6" + alias));
            }

            sender.sendMessage(Components.fromLegacy("&7Syntaxes:"));
            getSyntaxes().stream().forEach(syntax -> {
                String syntaxContent = "&7 - &f" + getName();

                for (Argument arg : syntax.getArguments())
                {
                    syntaxContent += (" " + arg.getId());

                    /**
                    switch (arg.getClass().getName())
                    {
                        case "ArgumentEntity":
                            syntaxContent += "<player>";
                            break;
                        case "ArgumentEnum":
                            String enumText = "<";
                            Iterator<String> enumIterator = ((ArgumentEnum) arg).entries().iterator();
                            while(enumIterator.hasNext())
                            {

                            }
                    }
                     */
                }

                sender.sendMessage(Components.fromLegacy(syntaxContent));
            });
        });
    }

    public AbstractCommand getParent()
    {
        return this.parent;
    }

    public abstract Permission permission();
}

File: src/main/kotlin/io/myosotisdev/minestom/command/AbstractCommand.kt
package io.myosotisdev.minestom.command

import io.myosotisdev.minestom.Minestom
import io.myosotisdev.minestom.permission.Permission
import io.myosotisdev.minestom.util.Commands
import io.myosotisdev.minestom.util.Components
import net.minestom.server.command.CommandSender
import net.minestom.server.command.ConsoleSender
import net.minestom.server.command.builder.Command
import net.minestom.server.command.builder.CommandContext
import net.minestom.server.command.builder.CommandExecutor
import net.minestom.server.command.builder.CommandSyntax
import net.minestom.server.command.builder.arguments.Argument

abstract class AbstractCommand(val parent: AbstractCommand?, name: String, vararg aliases: String?) : Command(name, *aliases)
{

    init
    {
        defaultExecutor = CommandExecutor { sender: CommandSender, context: CommandContext? ->
            if(Minestom.checkPermission(sender, permission()))
            {
                showUsages(sender)
            }
            else
                sender.sendMessage(Components.fromLegacy("&4You don't have enough permission to do that!"))
        }
    }

    fun addSyntax(executor: CommandExecutor, permission: Permission? = permission(), allowConsole: Boolean = false, vararg args: Argument<*>): MutableCollection<CommandSyntax>
    {
        return super.addSyntax({ sender, context ->
            if(sender is ConsoleSender && !allowConsole)
            {
                sender.sendMessage(Components.fromLegacy("&4This command could only be used by a player."))
                return@addSyntax
            }

            if(!Minestom.checkPermission(sender, permission))
            {
                sender.sendMessage(Components.fromLegacy("&4You don't have enough permission to do that!"))
                return@addSyntax
            }

            executor.apply(sender, context)
        }, *args)
    }

    fun addSyntax(executor: CommandExecutor, allowConsole: Boolean, vararg args: Argument<*>): MutableCollection<CommandSyntax>
    {
        return addSyntax(executor, permission(), allowConsole, *args)
    }

    override fun addSyntax(executor: CommandExecutor, vararg args: Argument<*>): MutableCollection<CommandSyntax>
    {
        return addSyntax(executor, permission(), false, *args)
    }

    fun showUsages(sender: CommandSender)
    {
        sender.sendMessage(Components.fromLegacy("&aCommand: &8[&6" + getName() + "&8]"))
        sender.sendMessage(Components.fromLegacy("&7Aliases:"))
        getAliases()?.forEach { alias ->
            sender.sendMessage(Components.fromLegacy("&7 - &6$alias"))
        }
        sender.sendMessage(Components.fromLegacy("&7Syntaxes:"))
        Commands.usages(this).stream().forEach { string ->
            sender.sendMessage(Components.fromLegacy("&7 - &f" + string))
        }
    }

    abstract fun permission(): Permission?
}